### PR TITLE
add usage based scheduling plugin

### PR DESCRIPTION
**Pull request by william-wang**
- The feature is for issue #1777 
- The design doc is #2023

### CODE REVIEW

File: pkg/scheduler/plugins/usage/usage.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+			case int:
+				val = float64(a)
+			case float64:
+				val = a
+			default:
+				klog.V(4).Infof("The threshold %v is an unknown type", a)
+			}
+			if strings.Contains(key, cpuUsageAvgPrefix) {
+				periodKey := strings.Replace(key, cpuUsageAvgPrefix, "", 1)
+				up.threshold.cpuUsageAvg[periodKey] = val
```
Comment (wpeng102) [inline pkg/scheduler/plugins/usage/usage.go:115]: it seems `periodKey` is `.5m`, it is correct?

File: pkg/scheduler/plugins/usage/usage.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+				return fmt.Errorf("plugin %s memory usage predicates failed %s", up.Name(), msg)
+			}
+		}
+		klog.V(4).Infof("Usage plugin filter for task %s/%s on node %s pass.", task.Namespace, task.Name, node.Name)
+		return nil
+	}
+
+	nodeOrderFn := func(task *api.TaskInfo, node *api.NodeInfo) (float64, error) {
+		score := 0.0
+		cpuUsage, exist := node.ResourceUsage.CPUUsageAvg[cpuUsageAvg5m]
```
Comment (wpeng102) [inline pkg/scheduler/plugins/usage/usage.go:149]: nodeOrderFn only consider cpu usage?

**Review by wpeng102 — COMMENTED** [review]
/lgtm

File: pkg/scheduler/plugins/usage/usage.go (hunk 2 above)
Comment (william-wang) [reply pkg/scheduler/plugins/usage/usage.go:149]: Yes. in this alpha version, only cpu is considered. In the future we will usie drf to consider the mem and cpu in pod request and node multiple dimension resource together.

File: pkg/scheduler/plugins/usage/usage.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package usage
+
+import (
+	"fmt"
```
Comment (Thor-wl) [inline pkg/scheduler/plugins/usage/usage.go:20]: Please sort packages.

File: pkg/scheduler/plugins/usage/usage.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+type usagePlugin struct {
+	pluginArguments framework.Arguments
+	weight          int
+	threshold       thresholdConfig
+}
+
+// New function returns usagePlugin object
+func New(args framework.Arguments) framework.Plugin {
+	usageWeight := 1
```
Comment (Thor-wl) [inline pkg/scheduler/plugins/usage/usage.go:62]: What does `usageWeight` mean? Where is the validation?

File: pkg/scheduler/plugins/usage/usage.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
+/*
```
Comment (Thor-wl) [inline pkg/scheduler/plugins/usage/usage.go:1]: Suggest adding UT for new plugins.

File: pkg/scheduler/api/node_info.go [hunk 6]
```diff
@@ -92,6 +93,25 @@ type NodeState struct {
 	Reason string
 }
 
+type NodeUsage struct {
```
[the remark this answers is not in the document]
Comment (william-wang) [reply pkg/scheduler/api/node_info.go:96]: OK, I will add it, thanks

File: pkg/scheduler/cache/cache.go [hunk 7]
```diff
@@ -47,6 +47,8 @@ import (
 	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
 	volumescheduling "k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding"
 
+	"github.com/prometheus/client_golang/api"
```
[the remark this answers is not in the document]
Comment (william-wang) [reply pkg/scheduler/cache/cache.go]: fixed

File: pkg/scheduler/cache/cache.go [hunk 8]
```diff
@@ -61,6 +63,15 @@ import (
 	"volcano.sh/volcano/pkg/scheduler/metrics"
 )
 
+const (
+	// record name of cpu average usage defined in prometheus rules
+	cpuUsageAvg = "cpu_usage_avg"
+	// record name of mem average usage defined in prometheus rules
+	memUsageAvg = "mem_usage_avg"
+	// default interval for sync data from metrics server
+	defaultMetricsInternal = 5
```
[the remark this answers is not in the document]
Comment (william-wang) [reply pkg/scheduler/cache/cache.go:73]: 5s, just added comments to clarify this.

File: pkg/scheduler/plugins/usage/usage.go (hunk 4 above)
Comment (william-wang) [reply pkg/scheduler/plugins/usage/usage.go:64]: it's the plugin weight. like binpack.weight

File: pkg/scheduler/plugins/usage/usage.go (hunk 5 above)
Comment (william-wang) [reply pkg/scheduler/plugins/usage/usage.go:1]: yes. will add it in another pr.

File: pkg/scheduler/cache/cache.go [hunk 9]
```diff
@@ -1155,3 +1174,79 @@ func (sc *SchedulerCache) recordPodGroupEvent(podGroup *schedulingapi.PodGroup,
...
+				}
+				if len(warnings) > 0 {
+					klog.V(3).Infof("Warning querying Prometheus: %v", warnings)
+				}
+
+				if err == nil {
+					rowValues := strings.Split(strings.TrimSpace(res.String()), "=>")
+					value := strings.Split(strings.TrimSpace(rowValues[1]), " ")
+					switch metric {
+					case "cpu_usage_avg":
```
[the remark this answers is not in the document]
Comment (william-wang) [reply pkg/scheduler/cache/cache.go]: updated.

File: pkg/scheduler/plugins/usage/usage.go (hunk 4 above)
Comment (Thor-wl) [reply pkg/scheduler/plugins/usage/usage.go:64]: Got it! That's reasonable. Sussget adding some comment to make it clear.

**Pull request merged**